### PR TITLE
[WIP] Change the recommendation logic

### LINKED ISSUE #65: [Feature][Impl] - Stack Analysis modified logic for recommendations

### PR DESCRIPTION
**Pull request by arunkumars08**
graphaggregator.py
1. Read user profile information as sent from api_v1.py's StackAnalysis Call
2. Add the logic to store the user profile information to Amazon S3 bucket
3. Modify the return object with the addition of user profile information to be able to use in recommender flow

recommender.py
1. Introduce a new execute flow for the modified logic of recommendation StackAnalysis
2. Setup a function to call Janus model and pass user profile information to get the user persona information
3. Use the user persona information and call the PGM function with the normalized manifest data to get the relevant packages
4. Call Gremlin with a new query by passing ecosystem and packages list received from PGM to get their version information

Fixes #65

### CODE REVIEW

File: cucoslib/workers/graphaggregator.py [hunk 1]
```diff
@@ -21,8 +25,25 @@ def _handle_external_deps(self, ecosystem, deps):
         versions = solver.solve(deps)
         return [{"package": k, "version": v} for k, v in versions.items()]
 
+    def store_in_bucket(self, content):
+        bucket_name = 'janus-json-data'
+        try:
+            email = content.get('email')
+            s3 = AmazonS3(bucket_name=bucket_name)
+            s3.connect()
```
Comment (fridex) [inline cucoslib/workers/graphaggregator.py]: All connections are handled by Selinon - see other parts of the project. An example could be:

```python
from selinon import StoragePool

s3 = StoragePool.get_connected_storage('your-storage-adapter-goes-here')
```
You will need to create your own S3 adapter that handles your logic and place it to `cucoslib/storages/` - see other adapters as an example. All configuration related pieces go to `cucoslib/dispatcher/nodes.yml` file accordingly, where you set up whether you want to encrypt, version files and other things.

You will also need to create custom methods such for storing user profile to encapsulate logic (do not place it here as bellow).

File: cucoslib/workers/graphaggregator.py [hunk 2]
```diff
@@ -9,6 +9,10 @@
 
 from cucoslib.workers.mercator import MercatorTask
 
+# AWS Related Imports
+from cucoslib.storages import AmazonS3
+# AWS Related Imports
+
```
Comment (fridex) [inline cucoslib/workers/graphaggregator.py]: Replace this with `from selinon import StoragePool` as described.

File: cucoslib/workers/graphaggregator.py [hunk 3]
```diff
@@ -21,8 +25,25 @@ def _handle_external_deps(self, ecosystem, deps):
         versions = solver.solve(deps)
         return [{"package": k, "version": v} for k, v in versions.items()]
 
+    def store_in_bucket(self, content):
+        bucket_name = 'janus-json-data'
```
Comment (fridex) [inline cucoslib/workers/graphaggregator.py]: This should go to config files.

File: cucoslib/workers/graphaggregator.py [hunk 4]
```diff
@@ -21,8 +25,25 @@ def _handle_external_deps(self, ecosystem, deps):
...
+        try:
+            email = content.get('email')
+            s3 = AmazonS3(bucket_name=bucket_name)
+            s3.connect()
+
+            folder = 'new_user_profile'
+            key = folder + '/' + email + '.json'
+            s3.store_dict(content, key)
+        except:
+            raise Exception('Amazon Exception')
```
Comment (fridex) [inline cucoslib/workers/graphaggregator.py]: You probably don't want to do this as these errors are tracked inside aggregated logs. Re-raising here does not add any additional info for exception.

File: cucoslib/workers/graphaggregator.py [hunk 5]
```diff
@@ -21,8 +25,25 @@ def _handle_external_deps(self, ecosystem, deps):
...
+
     def execute(self, arguments):
         self._strict_assert(arguments.get('manifest'))
+        self._strict_assert(arguments.get('user_profile'))
+
+        user_profile = arguments.get('user_profile')
```
Comment (fridex) [inline cucoslib/workers/graphaggregator.py]: You can now directly access `user_profile` without `get()` method as the assert bellow guarantees existence.

File: cucoslib/workers/recommender.py [hunk 6]
```diff
@@ -43,6 +43,7 @@ class GraphDB:
 
     def __init__(self):
         self._bayesian_graph_url = GREMLIN_SERVER_URL_REST
+        #self._bayesian_graph_url = "http://gremlin-http-thread-pool.dev.rdu2c.fabric8.io"
```
Comment (fridex) [inline cucoslib/workers/recommender.py]: Please remove such code in production (even commented out).

File: cucoslib/workers/recommender.py [hunk 7]
```diff
@@ -257,6 +258,31 @@ def get_reference_stacks_from_graph(self, list_packages):
 
         return list_ref_stacks
 
+    def get_version_information(self, input_list, ecosystem):
+        """Fetch the version information for each of the packages"""
+        input_packages = []
+        for package in input_list:
+            input_packages.append(package)
```
Comment (fridex) [inline cucoslib/workers/recommender.py]: Why there is not directly used `input_list`?

File: cucoslib/workers/recommender.py [hunk 8]
```diff
@@ -257,6 +258,31 @@ def get_reference_stacks_from_graph(self, list_packages):
...
+                'ecosystem': ecosystem,
+                'input_packages': input_packages
+            }
+        }
+        
+        #Query Gremlin with packages list to get their version information
+        gremlin_response = self.execute_gremlin_dsl(payload)
+        if gremlin_response is None:
+            return []
+
```
Comment (fridex) [inline cucoslib/workers/recommender.py]: No need to put 2 new lines there.

File: cucoslib/workers/recommender.py [hunk 9]
```diff
@@ -257,6 +258,31 @@ def get_reference_stacks_from_graph(self, list_packages):
...
+            }
+        }
+        
+        #Query Gremlin with packages list to get their version information
+        gremlin_response = self.execute_gremlin_dsl(payload)
+        if gremlin_response is None:
+            return []
+        
+
+        print(gremlin_response)
```
Comment (fridex) [inline cucoslib/workers/recommender.py]: A debug left-over?

File: cucoslib/workers/recommender.py [hunk 10]
```diff
@@ -431,11 +456,134 @@ def find_relative_similarity(self, input_stack, input_stack_vectors, filtered_re
 class RecommendationTask(BaseTask):
     _analysis_name = 'recommendation'
     description = 'Get Recommendation'
+    
+    def call_janus_model(self, user_profile):
+        """Calls the Janus model with the user profile information
+            to get the user_persona"""
+        _janus_end_point = "http://janus-demo-janus.dev.rdu2c.fabric8.io/api/v1/janus"
```
Comment (fridex) [inline cucoslib/workers/recommender.py]: This should be configurable via env variables.

File: cucoslib/workers/recommender.py [hunk 11]
```diff
@@ -431,11 +456,134 @@ def find_relative_similarity(self, input_stack, input_stack_vectors, filtered_re
...
+        try:
+            if user_profile is not None:
+                response = get_session_retry().post(_janus_end_point, json=user_profile)
+                if response.status_code != 200:
+                    _logger.error ("HTTP error {}. Error retrieving Janus data.".format(response.status_code))
```
Comment (fridex) [inline cucoslib/workers/recommender.py]: All tasks have logger ready in `self.log` property. Please use it as we do additional configuration to properly aggregate logs. Here and other places as well.

File: cucoslib/workers/recommender.py [hunk 12]
```diff
@@ -431,11 +456,134 @@ def find_relative_similarity(self, input_stack, input_stack_vectors, filtered_re
...
+                    return json_response
+            else:
+                _logger.info('User Profile information is not passed in the call, Quitting! Janus model\'s call')
+        except:
+            _logger.error ("Failed retrieving Janus data.")
+            return None
+
+    def call_pgm(self, payload):
+        """Calls the PGM model with the normalized manifest information to get the relevant packages"""
+        _pgm_end_point = "http://kronos-kattappa-0237.dev.rdu2c.fabric8.io/api/v1/kronos_score"
```
Comment (fridex) [inline cucoslib/workers/recommender.py]: Again, please make this configurable via env variables.

File: cucoslib/workers/recommender.py [hunk 13]
```diff
@@ -431,11 +456,134 @@ def find_relative_similarity(self, input_stack, input_stack_vectors, filtered_re
...
+        try:
+            if payload is not None:
+                print(_pgm_end_point)
+                print(type(payload))
```
Comment (fridex) [inline cucoslib/workers/recommender.py]: Debug leftover prints? Note you can use `self.log.debug` for debug purposes (these messages are turned off in production).

File: cucoslib/workers/recommender.py [hunk 14]
```diff
@@ -431,11 +456,134 @@ def find_relative_similarity(self, input_stack, input_stack_vectors, filtered_re
...
+                    _logger.error ("HTTP error {}. Error retrieving PGM data.".format(response.status_code))
+                    return None
+                else:
+                    json_response = response.json()
+                    return json_response
+            else:
+                _logger.info('Payload information is not passed in the call, Quitting! PGM\'s call')
+        except:
+            _logger.error ("Failed retrieving PGM data.")
+            return None
```
Comment (fridex) [inline cucoslib/workers/recommender.py]: Do we really want to proceed even if there were no data? Maybe marking task as failed will help us when we would like to look for these kind of errors in aggregated logs.

File: cucoslib/workers/recommender.py [hunk 15]
```diff
@@ -431,11 +456,134 @@ def find_relative_similarity(self, input_stack, input_stack_vectors, filtered_re
...
+        
 
     def execute(self, arguments=None):
         arguments = self.parent_task_result('GraphAggregatorTask')
         rs = RelativeSimilarity()
 
+        results = arguments.get('result')
```
Comment (fridex) [inline cucoslib/workers/recommender.py]: `arguments['result']`

File: cucoslib/workers/recommender.py [hunk 16]
```diff
@@ -431,11 +456,134 @@ def find_relative_similarity(self, input_stack, input_stack_vectors, filtered_re
...
+        user_profile = arguments.get('user_profile')
```
Comment (fridex) [inline cucoslib/workers/recommender.py]: `arguments['user_profile']`

File: cucoslib/workers/recommender.py [hunk 17]
```diff
@@ -431,11 +456,134 @@ def find_relative_similarity(self, input_stack, input_stack_vectors, filtered_re
...
+        inputTaskForPGM = []
+
+        for result in results:
+            details = result['details'][0]
+            resolved = details['_resolved']
+            self.log.info(result)
+            newArr = [r['package'] for r in resolved]
+            jsonObject = {
+                'ecosystem': details['ecosystem'],
+                'user_persona': "1",
```
Comment (fridex) [inline cucoslib/workers/recommender.py]: Why there is hard-coded `"1"`?

File: cucoslib/workers/recommender.py [hunk 18]
```diff
@@ -431,11 +456,134 @@ def find_relative_similarity(self, input_stack, input_stack_vectors, filtered_re
...
+                'package_list': newArr
+            }
+            self.log.info(jsonObject)
+            inputTaskForPGM.append(jsonObject)
```
Comment (fridex) [inline cucoslib/workers/recommender.py]: Please be consistent with code - we use snake case according to PEP8 in the code.

File: cucoslib/workers/recommender.py [hunk 19]
```diff
@@ -431,11 +456,134 @@ def find_relative_similarity(self, input_stack, input_stack_vectors, filtered_re
...
+            else:
+                if dictionary[package_name]['dependents_count'] < dep_count:
+                    dictionary[package_name]['version'] = version
+                    dictionary[package_name]['dependents_count'] = dep_count
+                    recommendations[package_name] = version
+        
+        return recommendations
+
+
+    def execute_old_flow(self, arguments=None):
```
Comment (fridex) [inline cucoslib/workers/recommender.py]: If the code is not used, please remove it. We don't want to maintain code that is dead.

File: cucoslib/workers/recommender.py (hunk 16 above)
Comment (jpopelka) [inline cucoslib/workers/recommender.py]: and add `self._strict_assert(arguments.get('user_profile'))` at the beginning of execute()

File: cucoslib/workers/recommender.py [hunk 20]
```diff
@@ -431,11 +456,134 @@ def find_relative_similarity(self, input_stack, input_stack_vectors, filtered_re
...
+        # Which is then used as an input along with a few others to hit PGM 
+        # Which gives us the response with list of packages and that is later on passed to Gremlin to get the version information for those packages
+        
+        janus_response = self.call_janus_model(user_profile)
+        inputTaskForPGM = []
+
+        for result in results:
+            details = result['details'][0]
+            resolved = details['_resolved']
+            self.log.info(result)
```
Comment (jpopelka) [inline cucoslib/workers/recommender.py]: If it's a debug left-over, then either remove it or turn it into `self.log.debug(result)`

File: cucoslib/workers/recommender.py [hunk 21]
```diff
@@ -431,11 +456,134 @@ def find_relative_similarity(self, input_stack, input_stack_vectors, filtered_re
...
+            newArr = [r['package'] for r in resolved]
+            jsonObject = {
+                'ecosystem': details['ecosystem'],
+                'user_persona': "1",
+                'package_list': newArr
+            }
+            self.log.info(jsonObject)
```
Comment (jpopelka) [inline cucoslib/workers/recommender.py]: If it's a debug left-over, then either remove it or turn it into `self.log.debug(json_object)`

File: cucoslib/workers/recommender.py [hunk 22]
```diff
@@ -431,11 +456,134 @@ def find_relative_similarity(self, input_stack, input_stack_vectors, filtered_re
...
+
+                for pkg, ver in reco_hash.items():
+                    companion_out_arr.append(ecosystem + ':' + pkg + ':' + ver)
+
+            return recommendation
+
+
+    def get_recommendations_hash(self, results):
+        dictionary = {}
+        highest_usage_hash = {}
```
Comment (jpopelka) [inline cucoslib/workers/recommender.py]: unused

File: cucoslib/workers/recommender.py [hunk 23]
```diff
@@ -431,11 +456,134 @@ def find_relative_similarity(self, input_stack, input_stack_vectors, filtered_re
...
+            else:
+                _logger.info('Payload information is not passed in the call, Quitting! PGM\'s call')
+        except:
+            _logger.error ("Failed retrieving PGM data.")
+            return None
+        
 
     def execute(self, arguments=None):
         arguments = self.parent_task_result('GraphAggregatorTask')
         rs = RelativeSimilarity()
```
Comment (jpopelka) [inline cucoslib/workers/recommender.py]: `rs` is unused

File: cucoslib/workers/recommender.py [hunk 24]
```diff
@@ -431,11 +456,134 @@ def find_relative_similarity(self, input_stack, input_stack_vectors, filtered_re
...
 
+        results = arguments.get('result')
+        user_profile = arguments.get('user_profile')
+
+        # Send the user_profile information to Janus model to get back the user_persona
+        # Which is then used as an input along with a few others to hit PGM 
+        # Which gives us the response with list of packages and that is later on passed to Gremlin to get the version information for those packages
+        
+        janus_response = self.call_janus_model(user_profile)
```
Comment (jpopelka) [inline cucoslib/workers/recommender.py]: `janus_response` is unused

File: cucoslib/workers/recommender.py [hunk 25]
```diff
@@ -431,11 +456,134 @@ def find_relative_similarity(self, input_stack, input_stack_vectors, filtered_re
...
+            recommendation = {
+                'analysis': {
+                    'companion_packages': [],
+                    'alternate_packages': []
+                }
+            }
+
+            for result in pgm_response:
+                ecosystem = result['ecosystem']
+                user_persona = result['user_persona']
```
Comment (jpopelka) [inline cucoslib/workers/recommender.py]: `user_persona` is unused

File: cucoslib/workers/recommender.py (hunk 7 above)
Comment (arunkumars08) [inline cucoslib/workers/recommender.py]: input_list is a dictionary and what I need here is just an array

Comment (arunkumars08) [inline cucoslib/workers/recommender.py]: @fridex ^

File: cucoslib/workers/recommender.py (hunk 8 above)
Comment (arunkumars08) [inline cucoslib/workers/recommender.py]: Could you please explain, I don't seem to understand this comment @fridex

File: cucoslib/workers/recommender.py (hunk 24 above)
Comment (arunkumars08) [inline cucoslib/workers/recommender.py]: It is used now. @jpopelka

File: cucoslib/workers/recommender.py (hunk 20 above)
Comment (arunkumars08) [inline cucoslib/workers/recommender.py]: Handled as info @jpopelka

File: cucoslib/workers/recommender.py (hunk 7 above)
Comment (jpopelka) [inline cucoslib/workers/recommender.py]: Then why there's `list` in its name ?

You can use list comprehension in any case:

```python
input_packages = [package for package in input_list]
```

File: cucoslib/workers/recommender.py (hunk 8 above)
Comment (jpopelka) [inline cucoslib/workers/recommender.py]: There are 2 blank lines between `return []` and (now removed) `print(gremlin_response)` and PEP 8 doesn't like it.

**Comment by jpopelka** [comment]
Also it would be better (for you) if you next time create a branch (`git checkout -b name-of-change`) for your change and request merging that branch rather than your `master`.
That way you can work on more than one change at the same time.

**Comment by arunkumars08** [comment]
@jpopelka Yeah yeah. That's the ideal way. This I just made in a hurry. Will take care of that.

File: cucoslib/workers/recommender.py (hunk 19 above)
Comment (arunkumars08) [inline cucoslib/workers/recommender.py]: Handled

File: cucoslib/workers/recommender.py (hunk 18 above)
Comment (arunkumars08) [inline cucoslib/workers/recommender.py]: Handled

File: cucoslib/workers/recommender.py (hunk 17 above)
Comment (arunkumars08) [inline cucoslib/workers/recommender.py]: Handled

File: cucoslib/workers/recommender.py (hunk 16 above)
Comment (arunkumars08) [inline cucoslib/workers/recommender.py]: Done

File: cucoslib/workers/recommender.py (hunk 15 above)
Comment (arunkumars08) [inline cucoslib/workers/recommender.py]: Done

File: cucoslib/workers/recommender.py (hunk 13 above)
Comment (arunkumars08) [inline cucoslib/workers/recommender.py]: Ya. Leftovers. Handled

File: cucoslib/workers/recommender.py (hunk 12 above)
Comment (arunkumars08) [inline cucoslib/workers/recommender.py]: Done

**Pull request closed without merging**